velox/dwio/parquet/writer/Writer.cpp

```diff
@@ -223,12 +255,35 @@ dwio::common::StripeProgress getStripeProgress(
  *
  * This method assumes each input `ColumnarBatch` have same schema.
  */
-void Writer::write(const VectorPtr& data) {
+void Writer::write(const VectorPtr& dataToWrite) {
+  VELOX_USER_CHECK(
+      dataToWrite->type()->equivalent(*schema_),
+      "The file schema type should be equal with the input rowvector type.");
+  auto data = std::dynamic_pointer_cast<RowVector>(dataToWrite);
+  VELOX_CHECK_NULL(data->nulls());
+
+  // VELOX_CHECK_EQ(
```

Comment:
@mbasmanova https://github.com/facebookincubator/velox/pull/6074#discussion_r1416730736 Do you mean to add these two checks?

Reply:
Hmm.... dataToWrite is a shared pointer to a vector. In general, vectors are immutable, but it is Ok to modify a vector it is hasn't been shared yet, e.g. there is only single reference. setType() is a mutation API and therefore can only be called if vector is not shared. Here, I don't think you can guarantee that dataToWrite is singly-referenced. Hence, you may need to implement a different solution. E.g. pass the schema deeper into the writer and remove reliance on the vector's type.

Reply:
@mbasmanova I believe I understand your point now. Instead of calling the `data->setType(schema)` API to change the vector's schema, we can simply pass the schema itself instead of the vector when calling vec->type() in subsequent steps. In other words, for the [exportToArrow ](https://github.com/facebookincubator/velox/blob/main/velox/dwio/parquet/writer/Writer.cpp#L231)function below, we should pass the schema down rather than passing the vector. Please correct me if I have misunderstood. Thank you.

Reply:
@JkSelf Yes, I think so.

Reply:
I thought we tried this approach before and we concluded it won't work.
See related discussion here https://github.com/facebookincubator/velox/pull/6074#discussion_r1323872385

Reply:
@majetideepak I was thinking about this discussion, but couldn't remember the details. Thank you for the pointer. This is all very strange and we need to dig a bit deeper. At a high level, we are writing data from multiple Velox vectors into a file. We want to use particular names for columns in the file. Why do we have to make all vectors have the same field names? Shouldn't it be possible to specify column names to use in the file separately from column names used in the vectors? The discussion above talks about encodings, but different vectors can have different encodings and encodings should be separate from the column names, no?

Reply:
@mbasmanova @majetideepak 
According to the comment in https://github.com/facebookincubator/velox/pull/6627#discussion_r1330592350, we should pass vector down for encoding purposes. However, we can also pass the schema and use it in `exportToArrow `if the `schema `is not empty. Otherwise, we can use `vector->type()`. What do you think?

Reply:
@mbasmanova The answer to your question is the Arrow Schema API.  As Jimmy explained here https://github.com/facebookincubator/velox/pull/6627#discussion_r1330592350, the arrow::schema is a combination of vector encoding and type. We cannot separate them.
As Jimmy suggested, if a custom schema is provided, we can probably use an empty vector to serve as a container and get arrow::schema. We can then invoke `FileWriter::Open()` inside the `parquet::Writer` constructor instead of `flush`. This should define the Parquet file schema as the custom schema provided.

@JkSelf We do not want to pass schema to exportToArrow since that will further complicate the ArrowBridge API. Can you try the empty container approach mentioned above?

Reply:
@majetideepak It seems like using ArrowSchema to specify file schema is a hack and we need a cleaner way. File schema should not be tied to encoding.

Reply:
@mbasmanova we are currently limited by ArrowBridge API.  We need ArrowSchema to Open a Parquet FileWriter
```
/// \brief Try to create an Arrow to Parquet file writer.
....
static ::arrow::Result<std::unique_ptr<FileWriter>> Open(
      const ::arrow::Schema& schema,
      MemoryPool* pool,
      std::shared_ptr<::arrow::io::OutputStream> sink,
      std::shared_ptr<WriterProperties> properties =
          default_writer_properties(),
      std::shared_ptr<ArrowWriterProperties> arrow_properties =
          default_arrow_writer_properties());
```

Now that Arrow Writer code is in Velox, there is a task to remove this ArrowBridge layer and integrate more deeply.  We won't need this hack then.

Reply:
@majetideepak Got it. Thank you for clarifying. Then we should be able to make an empty vector of the desired type and call arrowToSchema.

Reply:
@majetideepak @majetideepak 
In Arrow, when writing a `RecordBatch `to a file, it is necessary to ensure that the `schema `of the `RecordBatch `matches the `schema `of the file. Therefore, we still need to modify the schema of the vector in this case. The related code is [here](https://github.com/apache/arrow/blob/main/cpp/src/arrow/chunked_array.cc#L75).

Reply:
How do we convert Velox Vector to RecordBatch? Can we rename fields as part of that conversion to avoid copying data?

Reply:
@mbasmanova @majetideepak  It seems that we can calculate `arrowContext_->schema` based on the `schema_ `passed into the writer's constructor. Then, when calling `Write(vector)`, instead of calculating `arrowSchema `based on the vector, we can directly use `arrowContext_->schema` to construct the `RecordBatch`. Can you help to review again?